### DISCUSSION TITLE
[@types/papaparse] does not handle direct Node Readable Stream Parsing

### DISCUSSION THREAD

**Discussion by likwidoxigen** [DefinitelyTyped/DefinitelyTyped#60543] [Issues with a @types package] [answered]
I just upgraded from @types/papaparse  5.2.4 to 5.3.2. Based on my testing this issue has existed since the 5.3.0 release. 

I use the stream processing within node and pass a fs.ReadStream to the parse object a lot like this:

```
const streamedFile: fs.ReadStream = await Common.asyncGetReadStream(filePath);
papa.parse(streamedFile, {
                header: false,
                delimiter: "|",
                step: async (row, parser) => await this.updateDB(row, parser, contentType),
                complete: (results) => {
                    console.log(`Done processing ${String(filePath.path)}`);
                    resolve("done");
                },
                error: (error: ParseError) => {
                    return reject(error);
                },
            });
```

With the pre 5.3.0 types this works without issue. With the newer 5.3.0 types  the error reported is:
```
src/loader.ts(167,24): error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'ReadStream' is not assignable to parameter of type 'unique symbol'.
```
Looking at the types file and the tests it appears that the types only account for the .pipe streaming style and not the option of being passed in directly as noted  [in the papaparse readme.](https://github.com/mholt/PapaParse#papa-parse-for-node)

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
there was a change in the main method overload and how the params are typed, see discussion in the PR, if it is the same problem:
#56667

**Reply by likwidoxigen** [reply]
That discussion seems related for using the .pipe syntax style and does not cover directly passing in a readable stream to parse which is also valid.

**Comment by OpportunityLiu** [comment, marked as answer]
You should not mark `error` as of type `ParseError`, it is defined as `Error` in type definition.

See  https://github.com/DefinitelyTyped/DefinitelyTyped/blob/19b70b2c6cbd626b1f861d4eef0954b0bf3fa5a5/types/papaparse/index.d.ts#L301

In papaparse, you'll find that all error thrown by `ReadableStream` will be handled by `error` callback, so that `error` can be any type of `Error`s.

https://github.com/mholt/PapaParse/blob/e11ee26581d9f497087b4c3ee00c2a77645b81d2/papaparse.js#L835-L842

https://github.com/mholt/PapaParse/blob/e11ee26581d9f497087b4c3ee00c2a77645b81d2/papaparse.js#L883-L887

https://github.com/mholt/PapaParse/blob/e11ee26581d9f497087b4c3ee00c2a77645b81d2/papaparse.js#L570-L582

**Reply by OpportunityLiu** [reply]
`ParseError`s will not send to the `error` handler, but to `step`, `chunk`, `complete`, as a part of `ParseResult` or `ParseStepResult`.

**Reply by OpportunityLiu** [reply]
Basically the same thing happed here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56667#issuecomment-1015028208

**Reply by likwidoxigen** [reply]
Using a readable stream as the first argument to parse is valid but is not accepted by the types. It used to be however with the changes in the 5.3+ series it is no longer accepted.

It's not similar to this [#56667 (comment)](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56667#issuecomment-1015028208) where the syntax to use .pipe is discussed and the specifically linked comment is regarding an incorrectly used attribute causing the signature not to match.

[The option for NodeJS.ReadableStream is what I'm referring to.](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/147586e5a9a3bccea55da62bdd52436878786689/types/papaparse/index.d.ts#L23)

Thanks for the advice with the code, it's very much appreciated, I'll make adjustments.

**Comment by likwidoxigen** [comment]
Unless there's any additional clarification I'll create the issue from the discussion.

**Reply by OpportunityLiu** [reply]
I think I'd answered clearly.

Once more:

```diff
 const streamedFile: fs.ReadStream = await Common.asyncGetReadStream(filePath);
 papa.parse(streamedFile, {
                header: false,
                delimiter: "|",
                step: async (row, parser) => await this.updateDB(row, parser, contentType),
                complete: (results) => {
                    console.log(`Done processing ${String(filePath.path)}`);
                    resolve("done");
                },
-               error: (error: ParseError) => {
+               error: (error: Error) => {
                    return reject(error);
                },
            });
```

**Reply by likwidoxigen** [reply]
Apologies. In my back and forth between the code modifications and the two @type versions I clearly was lacking in diligence with my tests. 
Thank you for your patience, I'll do better in the future.
